modules/tensor_mechanics/src/materials/FiniteStrainHyperElasticViscoPlastic.C

```diff
@@ -0,0 +1,394 @@
+{
+  _stress[_qp].zero();
+
+  _fp[_qp].zero();
+  _fp[_qp].addIa(1.0);
+
+  _q[_qp].resize(_num_internal_var, 0.0);
+  _q_old[_qp].resize(_num_internal_var, 0.0);
+}
+
+void
+FiniteStrainHyperElasticViscoPlastic::computeQpStress()
+{
+  if (_t_step == 0) return;
+
+  RankTwoTensor delta_dfgrd = _deformation_gradient[_qp] - _deformation_gradient_old[_qp];
+  unsigned int num_substep = 1;
+  unsigned int substep_iter = 1;
+
+  _err_flag = true;
+  while (_err_flag)
+  {
+    preSolveQp();
+    _dt_substep = _dt/num_substep;
+
+    for (unsigned int istep = 0; istep < num_substep; ++istep)
+    {
+      _dfgrd_tmp = (istep+1) * delta_dfgrd/num_substep + _deformation_gradient_old[_qp];
+      solveQp();
+      if (_err_flag)
+      {
+        substep_iter++;
+        num_substep*=2;
+        break;
+      }
+    }
+
+    if (substep_iter > _max_substep_iter)
+      mooseError("Constitutive failure with substepping");
+
+    if (!_err_flag)
+      postSolveQp();
+  }
+}
+
+
+void
+FiniteStrainHyperElasticViscoPlastic::preSolveQp()
+{
+  _err_flag = false;
+  _fp_tmp_old_inv = _fp_old[_qp].inverse();
+
+  for (unsigned int i = 0; i < _num_internal_var; ++i)
+    _q_tmp_old[i] = _q_old[_qp][i];
+}
+
+void
+FiniteStrainHyperElasticViscoPlastic::solveQp()
+{
+  preSolveStress();
+  solveStress();
+  if ( _err_flag )
+    return;
+  postSolveStress();
+}
+
+void
+FiniteStrainHyperElasticViscoPlastic::postSolveQp()
+{
+  _stress[_qp] = _fe * _pk2_tmp * _fe.transpose() / _fe.det();
+  _fp[_qp] = _fp_tmp_inv.inverse();
+
+  for (unsigned int i = 0; i < _num_internal_var; ++i)
+    _q[_qp][i] = _q_tmp[i];
+}
+
+void
+FiniteStrainHyperElasticViscoPlastic::preSolveStress()
+{
+  for (unsigned int i = 0; i < _num_flow_rate_uos; ++i)
+    _flow_rate[i] = 0.0;
+
+  for (unsigned int i = 0; i < _num_internal_var; ++i)
+  {
+    _q_tmp[i] = _q_tmp_old[i];
+    _dq_dflowrate[i] = 1.0;
+  }
+
+  _fp_tmp_inv = _fp_tmp_old_inv;
+  _fe = _dfgrd_tmp * _fp_tmp_inv;
+}
+
+void
+FiniteStrainHyperElasticViscoPlastic::solveStress()
+{
+  Real resid0, rnorm;
+  unsigned int iter = 0;
+
+#ifdef DEBUG
+  std::vector<Real> rnormst(_maxiters+1),flowratest(_maxiters+1);
+#endif
+
+  computeFlowRateResidual();
+  if (_err_flag)
+    return;
+
+  rnorm = computeNorm(&_resid[0]);
+  resid0 = rnorm;
+
+#ifdef DEBUG
+  rnormst[iter] = rnorm;
+  flowratest[iter] = computeNorm(&_flow_rate[0]);
+#endif
+
+  while (rnorm > _resid_abs_tol && rnorm > _resid_rel_tol * resid0 && iter < _maxiters)
+  {
+    computeFlowRateJacobian();
+    updateFlowRate();
+    computeElasticPlasticDeformGrad();
+    updateInternalVar();
+    if (_err_flag)
+      return;
+    computeFlowRateResidual();
+    if (_err_flag)
+      return;
+
+    rnorm = computeNorm(&_resid[0]);
+    iter++;
+
+#ifdef DEBUG
+    rnormst[iter] = rnorm;
+    flowratest[iter] = computeNorm(&_flow_rate[0]);
+#endif
+  }
+
+  if (iter == _maxiters && rnorm > _resid_abs_tol && rnorm > _resid_rel_tol * resid0)
+    _err_flag = true;
```

Comment:
@cpritam - We recently added a new mooseException() capability. Does failing this solve cause the step to fail? If so, this should be re-written to call mooseException(). 

Also, While I'm thinking about it, you should go back and try to use mooseException() with that old plasticity model you built where you were setting random numbers in the tensor to force failure. Yes, we can FINALLY fail gracefully from within the solve.